src/AirshipBindings.NETStandard/AirshipBindings.NETStandard.Android/MessagePageRenderer.cs

```diff
@@ -0,0 +1,166 @@
+using Android.Views;
+using Android.Webkit;
+using Android.Widget;
+using UrbanAirship.NETStandard.MessageCenter;
+using UrbanAirship.RichPush;
+using UrbanAirship.Widget;
+using Xamarin.Forms.Platform.Android;
+
+namespace UrbanAirship.NETStandard.Android
+{
+    public class MessagePageRenderer : PageRenderer, RichPushInbox.IFetchMessagesCallback
+    {
+        private class MessageWebViewClient : UAWebViewClient
+        {
+            private MessagePageRenderer renderer;
+            private bool error = false;
+
+            public MessageWebViewClient(MessagePageRenderer messagePageRenderer)
+            {
+                renderer = messagePageRenderer;
+            }
+
+            public override void OnPageFinished(WebView view, string url)
+            {
+                base.OnPageFinished(view, url);
+
+                if (renderer.message == null)
+                {
+                    return;
+                }
+
+                if (error)
+                {
+                    renderer.messagePage.OnRendererLoadFailed(renderer.message.MessageId, false, MessageFailureStatus.LoadFailed);
+                    return;
+                }
+
+                renderer.message.MarkRead();
+                renderer.messagePage.OnRendererLoaded(renderer.message.MessageId);
+            }
+
+#pragma warning disable CS0672 // Member overrides obsolete member
+            public override void OnReceivedError(WebView view, [GeneratedEnum] ClientError errorCode, string description, string failingUrl)
+#pragma warning restore CS0672 // Member overrides obsolete member
+            {
+#pragma warning disable CS0618 // Type or member is obsolete
+                base.OnReceivedError(view, errorCode, description, failingUrl);
+#pragma warning restore CS0618 // Type or member is obsolete
+
+                if (renderer.message != null && failingUrl != null && failingUrl.Equals(renderer.message.MessageBodyUrl))
+                {
+                    error = true;
+                }
+            }
+
+            public override void OnClose(WebView webView)
+            {
+                base.OnClose(webView);
+                if (renderer.message != null)
+                {
+                    renderer.messagePage.OnRendererClosed(renderer.message.MessageId);
+                }
+            }
+        }
+
+        private RichPushMessage message;
+        private ICancelable fetchMessageRequest;
+        private UAWebView webView;
+        private MessagePage messagePage;
+        private WebViewClient webViewClient;
+        private string messageId;
+        private FrameLayout nativeView;
+
+        public MessagePageRenderer(Context context) : base(context)
+        {
+            webView = new UAWebView(Context);
+            webViewClient = new MessageWebViewClient(this);
+
+            var fl = new FrameLayout(Context);
+            fl.LayoutParameters = new LayoutParams(LayoutParams.MatchParent, LayoutParams.MatchParent);
```

Comment:
Might not actually be needed to wrap in a framelayout. I had to do that it react because of a react bug. Doesnt hurt though.

Reply:
gotcha